galsim/photon_array.py

```diff
@@ -179,3 +180,68 @@ def applyTo(self, photon_array):
         """Assign wavelengths to the photons sampled from the SED * Bandpass."""
         photon_array.wavelength = self.sed.sampleWavelength(
                 photon_array.size(), self.bandpass, rng=self.rng, npoints=self.npoints)
+
+
+class FRatioAngles(object):
+    """A surface-layer operator that assigns photon directions based on the f/ratio and
+    obscuration.
+
+    Assigns arrival directions at the focal plane for photons, drawing from a uniform
+    brightness distribution between the obscuration angle and the angle of the FOV defined
+    by the f/ratio of the telescope.  The angles are expressed in terms of slopes dx/dz
+    and dy/dz.
+
+    @param fratio           The f/ratio of the telescope (e.g. 1.2 for LSST)
+    @param obscuration      Linear dimension of central obscuration as fraction of aperture
+                            linear dimension. [0., 1.).  [default: 0.0]
+    @param rng              A random number generator to use or None, in which case an rng
+                            will be automatically constructed for you. [default: None]
+    """
+    def __init__(self, fratio, obscuration=0.0, rng=None):
+
+        if fratio < 0:
+            raise ValueError("The f-ratio must be positive.")
+        if obscuration < 0 or obscuration >= 1:
+            raise ValueError("The obscuration fraction must be between 0 and 1.")
+        if rng is None:
+            ud = galsim.UniformDeviate()
+        else:
+            ud = galsim.UniformDeviate(rng)
+
+        self.fratio = fratio
+        self.obscuration = obscuration
+        self.ud = ud
+
+
+    def applyTo(self, photon_array):
+        """Assign directions to the photons in photon_array."""
+
+        dxdz = photon_array.getDXDZArray()
+        dydz = photon_array.getDYDZArray()
+        n_photons = len(dxdz)
+
+        # The f/ratio is the ratio of the focal length to the diameter of the aperture of
+        # the telescope.  The angular radius of the field of view is defined by the 
+        # ratio of the radius of the aperture to the focal length
+        fov_angle = np.arctan(0.5 / self.fratio)  # radians
+        obscuration_angle = self.obscuration * fov_angle
```

Comment:
Thanks for this.  I had the mistaken impression that the obscuration fraction was defined as a fraction of the pupil angular radius but I see that in Mike's description it is a fraction of the aperture size.